pyprep/find_noisy_channels.py

```diff
@@ -212,209 +234,229 @@ def find_bad_by_nan_flat(self):
 
     def find_bad_by_deviation(self, deviation_threshold=5.0):
-        """Robust z-score of the robust standard deviation for each channel is calculated.
+        """Detect channels with abnormally high or low overall amplitudes.
 
-        Channels having a z-score greater than 5 are detected as bad.
+        A channel is considered "bad-by-deviation" if its amplitude deviates
+        considerably from the median channel amplitude, as calculated using a
+        robust Z-scoring method and the given deviation threshold.
+
+        Amplitude Z-scores are calculated using the formula
+        ``(channel_amplitude - median_amplitude) / amplitude_sd``, where
+        channel amplitudes are calculated using a robust outlier-resistant estimate
+        of the signals' standard deviations (IQR scaled to units of SD), and the
+        amplitude SD is the IQR-based SD of those amplitudes.
 
         Parameters
         ----------
-        deviation_threshold : float
-            z-score threshold above which channels will be labelled bad.
+        deviation_threshold : float, optional
+            The minimum absolute z-score of a channel for it to be considered
+            bad-by-deviation. Defaults to 5.0.
 
         """
-        deviation_channel_mask = [False] * (self.n_chans_original)
-        robust_channel_deviation = np.zeros(self.n_chans_original, dtype=np.float64)
-        channel_deviation = np.zeros(self.n_chans_new)
-        for i in range(0, self.n_chans_new):
-            channel_deviation[i] = 0.7413 * _mat_iqr(self.EEGData[i, :])
-        channel_deviationSD = 0.7413 * _mat_iqr(channel_deviation)
-        channel_deviationMedian = np.nanmedian(channel_deviation)
-        robust_channel_deviation[self.usable_idx] = np.divide(
-            np.subtract(channel_deviation, channel_deviationMedian), channel_deviationSD
-        )
-        for i in range(0, self.n_chans_original):
-            deviation_channel_mask[i] = abs(
-                robust_channel_deviation[i]
-            ) > deviation_threshold or np.isnan(robust_channel_deviation[i])
+        IQR_TO_SD = 0.7413  # Scales units of IQR to units of SD, assuming normality
+
+        # Get channel amplitudes and the median / robust SD of those amplitudes
+        chan_amplitudes = _mat_iqr(self.EEGData, axis=1) * IQR_TO_SD
+        amp_sd = _mat_iqr(chan_amplitudes) * IQR_TO_SD
+        amp_median = np.nanmedian(chan_amplitudes)
+
+        # Calculate robust Z-scores for the channel amplitudes
+        amplitude_zscore = np.zeros(self.n_chans_original)
+        amplitude_zscore[self.usable_idx] = (chan_amplitudes - amp_median) / amp_sd
+
+        # Flag channels with amplitudes that deviate excessively from the median
+        abnormal_amplitude = np.abs(amplitude_zscore) > deviation_threshold
+        deviation_channel_mask = np.isnan(amplitude_zscore) | abnormal_amplitude
+
+        # Update names of bad channels by excessive deviation & save additional info
```

Comment:
nice vectorizing here

Reply:
Thanks! Was very satisfying to flatten those for-loops throughout 🙂